File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void solve(vector<int>&arr, vector<int>&temp, int target, int ind) {
        if(target < 0) return;
        if(target == 0) {
            ans.push_back(temp);
            return;
        }
        for(int i=ind; i<arr.size(); i++) {
            temp.push_back(arr[i]);
            solve(arr, temp, target - arr[i], i);
            temp.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int>temp;
        solve(candidates, temp, target, 0);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>>ans = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer>temp = new ArrayList<>();
        Arrays.sort(candidates);
        solve(candidates, target, temp, 0);
        return ans;
    }
    private void solve(int[] arr, int target, List<Integer>temp, int cur) {
        if(target == 0) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        if(target < 0) return;
        for(int i=cur; i<arr.length; i++) {
            temp.add(arr[i]);
            solve(arr, target - arr[i], temp, i);
            temp.remove(temp.size() - 1);
        }
    }
}